routing/http/server/filters.go

```diff
@@ -0,0 +1,229 @@
+package server
+
+import (
+	"reflect"
+	"slices"
+	"strings"
+
+	"github.com/ipfs/boxo/routing/http/types"
+	"github.com/ipfs/boxo/routing/http/types/iter"
+	"github.com/multiformats/go-multiaddr"
+)
+
+// filters implements IPIP-0484
+
+func parseFilter(param string) []string {
+	if param == "" {
+		return nil
+	}
+	return strings.Split(strings.ToLower(param), ",")
+}
+
+// applyFiltersToIter applies the filters to the given iterator and returns a new iterator.
+//
+// The function iterates over the input iterator, applying the specified filters to each record.
+// It supports both positive and negative filters for both addresses and protocols.
+//
+// Parameters:
+// - recordsIter: An iterator of types.Record to be filtered.
+// - filterAddrs: A slice of strings representing the address filter criteria.
+// - filterProtocols: A slice of strings representing the protocol filter criteria.
+func applyFiltersToIter(recordsIter iter.ResultIter[types.Record], filterAddrs, filterProtocols []string) iter.ResultIter[types.Record] {
+	mappedIter := iter.Map(recordsIter, func(v iter.Result[types.Record]) iter.Result[types.Record] {
+		if v.Err != nil || v.Val == nil {
+			return v
+		}
+
+		switch v.Val.GetSchema() {
+		case types.SchemaPeer:
+			record, ok := v.Val.(*types.PeerRecord)
+			if !ok {
+				logger.Errorw("problem casting find providers record", "Schema", v.Val.GetSchema(), "Type", reflect.TypeOf(v).String())
+				// TODO: Do we want to let failed type assertions to pass through?
+				return v
```

Comment:
I think if something went wrong here, its safer to skip result – may be malformed.

```suggestion
				// drop failed type assertion
				return iter.Result[types.Record]{}
```